testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jpa/mockprovider/skipquerydetach/SFSB1.java

```diff
@@ -0,0 +1,116 @@
+public class SFSB1 {
+    @PersistenceUnit(unitName = "queryresultsaremanaged")
+    EntityManagerFactory emf;
+
+    @PersistenceContext(unitName = "queryresultsaremanaged")
+    EntityManager em;
+
+    @PersistenceContext(unitName = "queryresultsaredetached")
+    EntityManager emQueryResultsAreDetached;
+
+    /**
+     * verify that entitymanager.clear is not called after executing returned query.
+     *
+     * Return null for success, otherwise for test failure return fail message.
+     */
+    public String queryWithSkipQueryDetachEnabled() {
+        if(TestEntityManager.getInvocations().contains("close")) {
+            return "invalid state, entity manager has already been previously closed";
+        }
+        if(TestEntityManager.getInvocations().contains("clear")) {
+            return "invalid state, entity manager has already been previously cleared";
+        }
+        Query query =  em.createQuery("mock query");
+        if(TestEntityManager.getInvocations().contains("close")) {
+            return "invalid state, entity manager was closed before query was executed, " +
+                    "which means that 'wildfly.jpa.skipquerydetach=true' couldn't work, as the " +
+                    "persistence context was closed too early";
+        }
+        if(TestEntityManager.getInvocations().contains("clear")) {
+            return "invalid state, entity manager was cleared (detached) before query was executed, " +
+                    "which means that 'wildfly.jpa.skipquerydetach=true' didn't work";
+        }
+        query.getSingleResult();
+        if(TestEntityManager.getInvocations().contains("close")) {
+            return "invalid state, entity manager was closed before SFSB1 bean call completed, " +
+                    "which means that 'wildfly.jpa.skipquerydetach=true' couldn't work, as the " +
+                    "persistence context was closed too early";
+        }
+        if(TestEntityManager.getInvocations().contains("clear")) {
+            return "invalid state, entity manager was cleared (detached) before SFSB1 bean call completed, " +
+                    "which means that 'wildfly.jpa.skipquerydetach=true' didn't work";
+        }
+        return null;  // success
+    }
+
+    /**
+     * verify that entitymanager.clear is called after executing returned query.
+     *
+     * Return null for success, otherwise for test failure return fail message.
+     */
+    public String queryWithSkipQueryDetachDisabled() {
+        if(TestEntityManager.getInvocations().contains("close")) {
+            return "invalid state, entity manager has already been previously closed";
+        }
+        if(TestEntityManager.getInvocations().contains("clear")) {
+            return "invalid state, entity manager has already been previously cleared";
+        }
+        Query query =  emQueryResultsAreDetached.createQuery("mock query");
+        if(TestEntityManager.getInvocations().contains("clear")) {
```

Comment:
And one more thing.. As we test the same `TestPersistenceProvider` (with the same `adapterClass`) but two different instances (which only differ in `skipquerydetach` property) - is there a reason that we don't have `TestEntityManager.getInvocations().contains("close")` check?

Reply:
On line 92, we do have "close" check, which shouldn't be needed but if there is an unexpected bug in the mock persistence provider EntityManager implementation, anything is possible, so best to check for if close or clear had already been called, which we do (lines 92 + 95).  

Then, after the query is created, we check again if clear has been called on line 99.